### ISSUE TITLE
Support GitLab groups/subgroups when querying API for projects/issues

### ISSUE THREAD

**Issue by Nezteb** [super-productivity/super-productivity#2529]
## Problem Statement

I'm trying to set up the GitLab.com integration, and my current settings aren't working no matter what values I use. In this case, my company's GitLab org (we'll call it `company`) has a `backend-team` _subgroup_.
<img width="1300" alt="Screenshot 2023-03-21 at 10 57 46 PM" src="https://user-images.githubusercontent.com/3588798/226816306-08d7b54c-73dc-4285-b2e4-ce9096e37d29.png">

## :grey_question: Possible Solution

Regardless of what "source" I set, the app tries to call the `/projects/<group>/issues` endpoint. From what I can tell, the official GitLab API docs don't mention that specific endpoint: https://docs.gitlab.com/ee/api/rest/index.html

If you instead use the `/groups/<group>%2<subgroup>/projects` (the `%2` is important) endpoint to get a list of projects and then iterate over that list of project IDs, you could get the issues from the `/projects/<id>/issues` endpoint.

## :arrow_heading_up: Describe alternatives you've considered

I've also tried just using the top-level group for the company, but that doesn't seem to work either.

## :heavy_plus_sign: Additional context

Relevant existing/closed issues and PRs:
- https://github.com/johannesjo/super-productivity/issues/697
- https://github.com/johannesjo/super-productivity/pull/1560

**Comment by Blackclaws** [comment]
You need to either put a group path at the top and select: "Group" as source

Or you need to put a project path and select "Project" as source.

Both work fine for me even for subgroups.

You need to paste the full group/project path too, so if you have a company group called:

CompanyA

with a subgroup Teams you need to put

CompanyA/Teams/BackendTeam

Let me know if you have further issues.

**Comment by Nezteb** [comment]
Attempting this again:

### Subgroup
- I have a group/team in GitLab: https://gitlab.com/company/backend-team
- I enter `company/backend-team` into the text box
- I set the "Source" to "Group", save, wait for the app to pull issues, and the app calls:
	- https://gitlab.com/api/v4/projects/company%2Fbackend-team/issues
	- This 404s
- I set the "Source" to "All", save, wait for the app to pull issues, and the app still calls the same URL and 404s

### Top-level group

- I enter just `company` into the text box
- The app now calls:
	- https://gitlab.com/api/v4/projects/company/issues
- The call still 404s
	- I tried with both "Group" and "All" for the "Source" dropdown, same result

### Project

If I repeat all of these same but for a specific repo, say https://gitlab.com/company/mobile-app, everything works, because it's using the `/projects` endpoint.

### Summary

So, I think pulling issues for an entire group is broken. There is code for handling groups, but it doesn't seem to be working: https://github.com/johannesjo/super-productivity/blob/9efed167a9d2b89be4f2389d1b58fec39a68732e/src/app/features/issue/providers/gitlab/gitlab-api/gitlab-api.service.ts#L397-L401

**Comment by excalios** [comment]
This issue is still relevant when getting notes or the issue detail itself is still giving an error. I would love it if anyone could help with this issue or give hints. Personally, this is very disturbing. Since, sometimes not all of my issue is loaded and even if it does it always shows an error popup.

When trying to get issues by its iids it will be an error since it will hit `projects/{PROJECT}` meanwhile `{PROJECT}` is set based on group or even username when set into GROUP or ALL. When getting issue details it has the same behavior.

I'm currently trying to understand the codebase and the GitLab API itself. So far what I got is:

# Getting by iids

## GROUP
When getting by iids it should still hit the groups endpoint but it currently hits the projects endpoint since we passed projects arguments from the GitLab config. Which is the group name, not the project name. I tried to fix it by changing the `_apiLink` function inside `gitlab_api.service.ts` specifically this part:
```ts
    if (
      (project !== projectConfig.project && projectConfig.project !== 'group') ||
      projectConfig.source === 'project'
    ) {
      apiURL += 'projects/' + projectURL;
    } else if (projectConfig.source === 'group') {
      apiURL += 'groups/' + projectURL;
    }
```
I checked whether the project that is passed is the same as the config and if it's not group it goes to group. Changing it into a switch and handling it inside switch case might be better.

## ALL
I haven't checked this but with the fix above it will try to hit projects with our username that is set

# Get issue detail

The issue is the same as it doesn't hit the proper projects, I tried to get from single issues endpoint but it's administrator only. I'm thinking it should pass the proper project id so the `getProject` function can return the project id for that specific issue. But, I haven't been able to find the specific part of the code that handles this logic.

**Comment by excalios** [comment]
is it possible to open this issue since it's still relevant? @johannesjo

**Comment by kronehero** [comment]
In version 10.0.11 on Windows 11, I can successfully access the issues for all group projects using the GitLab integration. I still can't use it because on the second automatic pull all issues are added again, so I get duplicates of issues I already have in my list. This doesn't happen if I use "Project" as source instead of "Group" in the integration settings.

In my eyes this problem is part of the GitHub issue and I just want to mention it here. Please let me know if it's not a problem at all and can be fixed by changing the configuration.

This is what my GitLab integration setup looks like:
<img src="https://github.com/user-attachments/assets/f2f115b6-5926-4db5-a485-20dd66e959b2" width="350">

**Comment by cbergmann** [comment]
Hi, after some month I tried the gitlab integration again. I found out that currently the "source" option is no longer available. I still have the requirement to search for issues on a group or instance level. Adding multiple Providers is not a option because there are many projects and they also change often.

I therefore propose the following change:

1. add a new Option for the gitlab issue provider "search scope" which has the options "project", "group", "instance".
2. The default for existing and new configurations is "project". This does not change the current function.
3. Depending on the selected option the search requests use the "/projects/:id/issues", "/groups/:id/issues" or "/issues" endpoint respectively.
4. For the ':id' in the "/groups/:id/issues" endpoint the "user name/project" config option is split by "/" and all but the last parts are used as ':id'.
5. Create and update requests always use the respective "/projects/:id/issues" endpoints.

I will look into this in the next days and see if I can create a small merge request. Any hints or improvements on this proposal are welcome. I have not looked into the datamodel if we need to change it to save the project of the issues to make the requests for "/projects/:id/issues/:id" etc. possible in this scheme as we could have issues from multiple project when searching for "group" or "instance" issues. Tips on this are also very welcome.

Kind regards
Clemens

**Comment by cbergmann** [comment]
Hi, this would be my proposal for the instance/group issue search feature: [!3901](https://github.com/johannesjo/super-productivity/pull/3901)

While testing I found that the submission of spent time seems to not be working also. I might look into this if I could find time.

**Comment by haydenseitz** [comment]
Is syncing GitLab issues from all projects in a group no longer available? I reviewed this issue and similar ones and found [a3b2675 ](https://github.com/johannesjo/super-productivity/commit/a3b2675) that looks like it removed the "Source" option to select either project or group.

**Comment by n-maym** [comment]
Hi!

I found this tool and was really happy to see it has a GitLab integration. However, it’s unfortunately not usable for me because I need the integration to support groups and subgroups.

In our setup, there are many subgroups, and each one has its own project where issues are tracked. I can’t use the tool efficiently if I have to create a new task for each new issue manually, and I also can’t sync issues properly when I can’t include all projects within a specific group path.

I assumed this feature existed by default, since it’s hard to imagine working with only a single project containing all issues. I understand this is open source and I might not be the target user, but without this functionality, it’s unusable for my workflow.

I’ve also noticed that several other issues have been opened requesting the same feature, so it seems I’m not the only one who needs it.

Or maybe I just can't find the correct configuration but I looked very hard :)

TL;DR: The software isn’t usable for me without group-based GitLab issue import.

Self-Hosted Gitlab EE.
Oversight over many Groups/Subgroups with own Issue Tracking Projects, ~100 (not all actively used all the time)

Thanks!

**Comment by cbergmann** [comment]
Hi, I have a similar setup. One CE Installation with multiple Groups/Subgroups containing multiple Projects with issues.

This setup seems to work but the configuration is not very obvious.

1. create a personal access token.
2. add gitlab-configuration with the following settings:
    1. username/project: '..' (only two dots) This will force Super-Productivity to use the global issues Search endpoint.
    2. Access Token: your access token
    3. custom gitlab issues URL: The Base URL of your installation 'https://gitlab.example.com'
    4. I would also suggest adding a filter. For example `state=opened` to not import old issues.

My setup is a little more elaborate. I have two Gitlab Connections configured. One of them has "auto import" enabled and `&my_reaction_emoji=eyes` added to the filter. I also have set up a Gitlab Trigger and Pipeline that automatically set my reaction to `eyes` whenever I am set as assignee on an issue. That way I automatically get my assigned issues in superproductivity and can add them to superproductivity by reacting in gitlab. If I need to add other issues I can always search in the other gitlab connection

**Comment by n-maym** [comment]
Thank you very much!!!

Link works and I can now fetch Issues.

Now I just need a way to differentiate between some administrative groups and some technical groups, so that issues are imported into the correct SuperProductivity project automatically, without having to create each task manually from the search function.

I did some testing but can't make this work with multiple Paths.
Is there now maybe something like ../group/subgroup/subgroup1 and then a second connector for ../group/subgroup/subgroup2 so that I can auto import to a specific project?

Additional note: this breaks the link provided in the task which should open the issue. But I could live with that.

**Comment by cbergmann** [comment]
The link works for me. Can you give me a (redacted) example of the generated link?

**Comment by n-maym** [comment]
I did some testing again and seperating it by (already existing) labels works.

This works for me for now. Hopefully the '..' trick will persist for all time. :)

Also the link in the summary of the imported issue (additional info) works, only the method right-click on task and "open in browser" is broken this way.

As for your question.
Via right-click on task and "open in browser":
https://gitlabdomain.redacted/-/issues/27

Via link in summary of additional info for that task:
https://gitlabdomain.redacted/group/subgroup1/subgroup2/subgroup3/project1/-/issues/27

Thanks again!

**Comment by johannesjo** [comment]
Should we reopen this?

**Comment by n-maym** [comment]
For my specific use case with nested groups and multiple projects the '..' solution works like a charm. Please keep it forever.

As for the issue with the links: I personally could live with that because I now got used to just click on the link in the summary instead the right-click menu.
I guess this could be fixed in a dedicated issue because in my opinion it has nothing to do with the group/subgroups request in this issue.

If you need information or I should test something (which doesn't break my current setup) I would happily do that if it helps :)

**Comment by Dustin1358** [comment]
We were using the '..' solution and was happy with that, but apparently you cannot set it up in the UI anymore. My existing configuration is still working but when I want to set up a new one it cannot save it as '..' is an invalid input.
